src/tests/p11test/p11test_case_common.c

```diff
@@ -413,9 +413,11 @@ int callback_public_keys(test_certs_t *objects,
 			RSA *rsa = RSA_new();
 			if (RSA_set0_key(rsa, n, e, NULL) != 1 ||
 			    EVP_PKEY_set1_RSA(o->key, rsa) != 1) {
+				RSA_free(rsa);
 				fail_msg("Unable to set key params");
 				return -1;
 			}
+			RSA_free(rsa);
```

Comment:
When `EVP_PKEY_set1_RSA` calls internally `EVP_PKEY_assign_RSA`, I think it should not be freed here (but I may not understand the [documentation](https://www.openssl.org/docs/man3.0/man3/EVP_PKEY_assign_RSA.html) correctly).

Reply:
The `*set1*`/`*get1*` functions set/get values by incrementing/decrementing their internal reference counter (in comparison to the `*set0*`/`*get0*` functions which hand over the ownership. Unfortunately documentation is not much helpful here, but source code helps noticing the difference between these two:
https://github.com/openssl/openssl/blob/b77826877be3bdd56e3e86887cb78ea010db90be/crypto/evp/p_legacy.c#L27-L31
Indeed, using `EVP_PKEY_assign_RSA()` directly might be simpler as it will help us avoid this free, but then I would have to break the condition to capture `RSA_set0_key()` failure separately (and free the RSA) and not free it in case of success of the `EVP_PKEY_set1_RSA()`. But I see that the EVP and RSA allocations are not checked so I will rewrite it anyway.

Reply:
Alright, thanks for clarification.